src/root.zig

```diff
@@ -508,9 +539,98 @@ pub const AsciiOptions = struct {
     }
 };
 
+pub const UnicodeOptions = struct {
+    const UnicodeScores = Scores(i32);
+
+    pub const TypeOfCharacter: type = u21;
+
+    case_sensitive: bool = true,
+    case_penalize: bool = false,
+    // treat spaces as wildcards for any kind of boundary
+    // i.e. match with any `[^a-z,A-Z,0-9]`
+    wildcard_spaces: bool = false,
+
+    penalty_case_mistmatch: i32 = -2,
+
+    /// Don't forget the allocator !!!
+    allocator: Allocator = undefined,
+
+    fn convertString(a: *const UnicodeOptions, string: []const u8) []const TypeOfCharacter {
+        var norm_data: Normalize.NormData = undefined;
+        Normalize.NormData.init(&norm_data, a.allocator) catch @panic("Cannot normalize string");
+        defer norm_data.deinit();
+
+        const n = Normalize{ .norm_data = &norm_data };
+
+        const nfc_result = n.nfc(a.allocator, string) catch @panic("Cannot normalize string");
+        defer nfc_result.deinit();
+
+        var iter = code_point.Iterator{ .bytes = nfc_result.slice };
+
+        var converted_string = std.ArrayList(TypeOfCharacter).init(a.allocator);
+        defer converted_string.deinit();
+
+        while (iter.next()) |c| {
+            converted_string.append(c.code) catch @panic("Memory error");
+        }
+        return converted_string.toOwnedSlice() catch @panic("Memory error");
+    }
+
+    fn eqlFunc(a: *const UnicodeOptions, h: u21, n: u21) bool {
+        const gcd = GenCatData.init(a.allocator) catch @panic("Memory error");
+        defer gcd.deinit();
+        if (gcd.isSeparator(n) and a.wildcard_spaces) {
+            if (gcd.isLetter(h) or gcd.isNumber(h) or gcd.isSymbol(h)) {
+                return true;
+            } else {
+                return false;
+            }
+        } else if (!a.case_sensitive) {
+            const cd = CaseData.init(a.allocator) catch @panic("Memory error");
+            defer cd.deinit();
+            return cd.toLower(h) == cd.toLower(n);
+        } else {
+            return h == n;
+        }
+    }
+
+    fn scoreFunc(
+        a: *const UnicodeOptions,
+        comptime scores: UnicodeScores,
+        h: u21,
+        n: u21,
+    ) ?i32 {
+        if (!a.eqlFunc(h, n)) return null;
+
+        if (a.case_penalize and (h != n)) {
+            return scores.score_match + a.penalty_case_mistmatch;
+        }
+        return scores.score_match;
+    }
+
+    fn bonusFunc(
+        self: *const UnicodeOptions,
+        comptime scores: UnicodeScores,
+        h: u21,
+        n: u21,
+    ) i32 {
+        const p = CharacterType.fromUnicode(h, self.allocator);
+        const c = CharacterType.fromUnicode(n, self.allocator);
+
+        return switch (p.roleNextTo(c)) {
+            .Head => scores.bonus_head,
+            .Camel => scores.bonus_camel,
+            .Break => scores.bonus_break,
+            .Tail => scores.bonus_tail,
+        };
+    }
+};
```

Comment:
These are both essentically identical to the Ascii functions which suggests we should pull them out so we only maintain the implementation once.